### PR TITLE
try to fix compile error in DeferredOrOffset

### PR DESCRIPTION
**Pull request by maxiee**
Problem Description:

I have created a demo project [flutter_eval_gallery](https://github.com/maxiee/flutter_eval_gallery) to experience flutter_eval. The project consists of two nested subprojects, the internal project is template which will be compiled into EVC bytecode by the dart_eval compiler. The external project in the root directory is the execution project. At runtime, the external project loads the template bytecode as assets and dynamically executes the template through the dart_eval runtime implementation.

The problem I'm having is that compiling the template with dart_eval compile is reporting errors.

Specifically, I have two computers, one running Windows 11 and one running macOS 11.6.8. The computer runs successfully on Windows, but the compilation of the template fails on the macOS computer.

Environment for both computers:

* Windows: Flutter 3.7.3, Dart 2.19.2, dart_eval 0.5.6, flutter_eval 0.5.4
* macOS: Flutter 3.7.3, Dart 2.19.2, dart_eval 0.5.6, flutter_eval 0.5.4

The above environment versions are the same for both computers. I'm trying to figure out why the macOS device fails to compile while the Windows device does.

macOS error log:

```
# flutter_eval_gallery % cd template
# template % dart_eval compile -o a.evc
Loading files...
Found binding file: .dart_eval/bindings/flutter_eval.json
Compiling package template...
Unhandled exception:
Null check operator used on a null value
#0      OffsetTracker.apply.<anonymous closure> (package:dart_eval/src/eval/compiler/offset_tracker.dart:19:95)
#1      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#2      OffsetTracker.apply (package:dart_eval/src/eval/compiler/offset_tracker.dart:16:22)
#3      Compiler.compileSources (package:dart_eval/src/eval/compiler/compiler.dart:380:27)
#4      Compiler.compile (package:dart_eval/src/eval/compiler/compiler.dart:131:12)
#5      main (file:///Users/maxiee/.pub-cache/hosted/pub.flutter-io.cn/dart_eval-0.5.6/bin/dart_eval.dart:125:36)
#6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
#7      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:192:26)
```

The apply method of OffsetTracker throws an exception.  The [source](https://github.com/ethanblake4/dart_eval/blob/ce0b32c4f9afd27201e88bde5d1c61c33b5eeb0f/lib/src/eval/compiler/offset_tracker.dart#L19) is as below:

```dart
List<EvcOp> apply(List<EvcOp> source) {
  _deferredOffsets.forEach((pos, offset) {
    final op = source[pos];
    if (op is Call) {
      final resolvedOffset = context.topLevelDeclarationPositions[offset.file!]![offset.name!]!;
      final newOp = Call.make(resolvedOffset);
      source[pos] = newOp;
    }
  });
  return source;
}
```

Where: `offset.name` is null, so `offset.name!` reports an error with the `Null check operator used on a null value`.

When I think about the cause of the problem, it happened when DeferredOrOffset was constructed, and the name was passed in as null during construction.

I found that there are many places where DeferredOrOffset is constructed, and there are many scenarios where it can be used, but the one that caused the macOS compilation to fail was one of them. With some debugging, I was able to pinpoint the location of the DeferredOrOffset construct that was causing the failure: the getValue() method of the IdentifierReference, located at [L301](https://github.com/ethanblake4/dart_eval/blob/a597622c23aa2cbb0f6846ff53b377519aa11fa0/lib/src/eval/compiler/reference.dart#L301):

```dart
if (decl is! FunctionDeclaration && decl is! ConstructorDeclaration) {
  decl as ClassDeclaration;

  final returnType = TypeRef.lookupClassDeclaration(ctx, _decl.sourceLib, decl);
  final DeferredOrOffset offset;

  if (ctx.topLevelDeclarationPositions[_decl.sourceLib]?.containsKey('$name.') ?? false) {
    offset =
        DeferredOrOffset(file: _decl.sourceLib, offset: ctx.topLevelDeclarationPositions[ctx.library]!['$name.']);
  } else {
    offset = DeferredOrOffset(file: _decl.sourceLib, name: '$name.');
  }

  return Variable(-1, EvalTypes.typeType,
      concreteTypes: [returnType], methodOffset: offset, methodReturnType: AlwaysReturnType(returnType, false));
}
```

In actual debugging, the name that causes the compile error to be reported as 'SubPage.' I found that it goes to the else branch on Windows devices and the if branch on macOS devices. This is an observed phenomenon, but the reason behind it is not well understood.

When the if branch is entered on macOS devices, the name parameter is not passed to the DeferredOrOffset constructor, resulting in an error when applying the OffsetTracker in the previous analysis.

I found that the name parameter is fully available in the if branch, so I tried to add it to the constructor. After adding it, macOS compiles and runs normally, and Windows still runs successfully.

Here's my fix for this PR, and the thinking behind it.

Since I do not yet fully understand how the project works and am still extrapolating more based on phenomena, my fix may cause potential problems. Please also discuss and analyze it together.

I tried to think further about the cause, and it looks like the `ctx.topLevelDeclarationPositions` generation process is different for the two computers. I have made an attempt.

When the `getValue()` of `IdentifierReference` is executed to 'SubPage.', the value of `ctx.topLevelDeclarationPositions` is printed, and the corresponding code is as follows.

```dart
if (decl is! FunctionDeclaration && decl is! ConstructorDeclaration) {
  decl as ClassDeclaration;

  final returnType = TypeRef.lookupClassDeclaration(ctx, _decl.sourceLib, decl);
  final DeferredOrOffset offset;

  // added by me
  if (name == "SubPage") {
    print(ctx.topLevelDeclarationPositions);
  }

  if (ctx.topLevelDeclarationPositions[_decl.sourceLib]?.containsKey('$name.') ?? false) {
    offset =
        DeferredOrOffset(file: _decl.sourceLib, offset: ctx.topLevelDeclarationPositions[ctx.library]!['$name.']);
  } else {
    offset = DeferredOrOffset(file: _decl.sourceLib, name: '$name.');
  }

  return Variable(-1, EvalTypes.typeType,
      concreteTypes: [returnType], methodOffset: offset, methodReturnType: AlwaysReturnType(returnType, false));
}
```

Windows:

```
{0: {}, 1: {}, 2: {}, 3: {}, 4: {}, 5: {}, 6: {}, 7: {}, 8: {}, 9: {}, 10: {}, 11: {Colors._: 34184}, 12: {}, 13: {}, 14: {}, 15: {}, 16: {}, 17: {}, 18: {}, 19: {}, 20: {}, 21: {}, 22: {}, 23: {}, 24: {}, 25: {}, 26: {}, 27: {}, 28: {}, 29: {}, 30: {Curves._: 34188}, 31: {}, 32: {}, 33: {}, 34: {}, 35: {}, 36: {}, 37: {}, 38: {}, 39: {}, 40: {}, 41: {}, 42: {}, 43: {}}
```

macOS:

```
{0: {}, 1: {}, 2: {}, 3: {}, 4: {}, 5: {}, 6: {}, 7: {}, 8: {}, 9: {}, 10: {}, 11: {Colors._: 34184}, 12: {}, 13: {}, 14: {}, 15: {}, 16: {}, 17: {}, 18: {}, 19: {}, 20: {}, 21: {}, 22: {}, 23: {}, 24: {}, 25: {}, 26: {}, 27: {}, 28: {}, 29: {}, 30: {Curves._: 34188}, 31: {}, 32: {}, 33: {}, 34: {}, 35: {}, 36: {}, 37: {}, 38: {}, 39: {}, 40: {}, 41: {}, 42: {}, 43: {SubPage.: 34236}, 44: {StatelessWidgetPage.: 34286}, 45: {StatefulWidgetPage.: 34297, _StatefulWidgetPageState.: 34366}, 46: {ViewModel.: 34386, ChangeNotifierPage.: 34400, _ChangeNotifierPageState.: 34545}, 47: {}}
```

They are indeed not quite the same.

### CODE REVIEW

**Comment by ethanblake4** [comment]
Wow, that's a really interesting issue. Thanks for the hard work.

First impression: the difference in OS is probably caused just because macOS and Windows list the files from the project in a different order when calling `listSync()` [here](https://github.com/ethanblake4/dart_eval/blob/2c78838d900f32c2c790fdfbba13f265ac2e8c1a/bin/dart_eval.dart#L98). Therefore, the files would be compiled in a different order in the compiler. But I have not tested that.

In any case, this is unfortunately not the correct fix. `DeferredOrOffset` is a reference to the position of a function in the EVC bytecode, e.g. main() might start at bytecode 13. However, sometimes you need to reference a function that hasn't yet been generated. It is supposed to have either the `offset` or the `name` property, but not both - `offset` if the function is already generated, and `name` if it hasn't. If you specify the name, the `OffsetTracker` will use the name later to resolve the `offset` after all the functions are generated. It doesn't make sense to specify both the offset _and_ the name.

My first guess was that a DeferredOrOffset that has an offset is getting added to the `OffsetTracker`, which is only supposed to be used for a DeferredOrOffset that has a name. After looking through the code the only place this is possible is [here](https://github.com/ethanblake4/dart_eval/blob/2c78838d900f32c2c790fdfbba13f265ac2e8c1a/lib/src/eval/compiler/expression/funcexpr_invocation.dart#L100), since every other call to addOffset() checks to make sure the offset is null. However, that code gets an offset from `getStaticDispatch()` not `getValue()` like your PR - unless you actually meant to correct [reference.dart#L385](https://github.com/ethanblake4/dart_eval/blob/2c78838d900f32c2c790fdfbba13f265ac2e8c1a/lib/src/eval/compiler/reference.dart#L385)? - It's the exact same code and in the same file, just copied and pasted later, so it would be easy to confuse.

Failing that, it's possible that somehow a DeferredOrOffset is being created with _both_ a null offset and name. You could check this by adding the following to the DeferredOrOffset constructor:
```dart
: assert(offset != null || name != null);
```

If needed I can dig out my old Macbook later this week and investigate, but if you can figure out anything that would be great. The question isn't "why is the nameless DeferredOrOffset being created" but "why is it being added to the OffsetTracker."

**Comment by maxiee** [comment]
Thank you for the detailed explanation, I gained a lot, thanks!

This isn't an urgent matter, but I find it very interesting and plan to continue exploring it.

You may not need to dig out your macbook anymore, I found a way to reproduce the problem under Windows too 😄 (or maybe GNU/Linux as well)

The master branch of flutter_eval_gallery is the branch that failed to compile.

After understanding how DeferredOrOffset works, my PR is indeed not the right fix.

**DeferredOrOffset's name and offset are both null**

As in your analysis, there is a case where both offset and name in DeferredOrOffset are null. Insert the following code into the DeferredOrOffset as you have given:

```dart
class DeferredOrOffset {
  DeferredOrOffset({this.offset, this.file, this.name, this.className, this.methodType, this.targetScopeFrameOffset}) : assert(offset != null || name != null) {
    if (offset == null && name == null) {
      throw "unexpected";
    }
  }
```

Explanation: I use `dart ~/Code/dart_eval/bin/dart_eval.dart compile -o . /assets/bundle.evc` to implement direct compilation with dart_eval source code. I found that in this case, assert could not break the program, so I added code to the constructor that throws an exception.

The call stack at the time of the problem was successfully captured as follows:

```
Unhandled exception:
unexpected
#0      new DeferredOrOffset (package:dart_eval/src/eval/compiler/offset_tracker.dart:36:7)
#1      IdentifierReference.getValue (package:dart_eval/src/eval/compiler/reference.dart:301:13)
#2      compileIdentifier (package:dart_eval/src/eval/compiler/expression/identifier.dart:13:48)
#3      compileInstanceCreation (package:dart_eval/src/eval/compiler/expression/instance_creation.dart:18:21)
#4      compileExpression (package:dart_eval/src/eval/compiler/expression/expression.dart:53:12)
#5      compileReturn (package:dart_eval/src/eval/compiler/statement/return.dart:19:17)
#6      compileStatement (package:dart_eval/src/eval/compiler/statement/statement.dart:24:12)
#7      compileIfStatement.<anonymous closure> (package:dart_eval/src/eval/compiler/statement/if.dart:14:47)
#8      macroBranch (package:dart_eval/src/eval/compiler/macros/branch.dart:21:32)
#9      compileIfStatement (package:dart_eval/src/eval/compiler/statement/if.dart:10:10)
#10     compileStatement (package:dart_eval/src/eval/compiler/statement/statement.dart:32:12)
#11     compileBlock (package:dart_eval/src/eval/compiler/statement/block.dart:16:20)
#12     compileMethodDeclaration (package:dart_eval/src/eval/compiler/declaration/method.dart:54:14)
#13     compileDeclaration (package:dart_eval/src/eval/compiler/declaration/declaration.dart:18:12)
#14     compileClassDeclaration (package:dart_eval/src/eval/compiler/declaration/class.dart:30:5)
#15     compileDeclaration (package:dart_eval/src/eval/compiler/declaration/declaration.dart:16:5)
#16     Compiler.compileSources.<anonymous closure>.<anonymous closure> (package:dart_eval/src/eval/compiler/compiler.dart:380:11)
#17     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#18     Compiler.compileSources.<anonymous closure> (package:dart_eval/src/eval/compiler/compiler.dart:369:15)
#19     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#20     Compiler.compileSources (package:dart_eval/src/eval/compiler/compiler.dart:366:32)
#21     Compiler.compile (package:dart_eval/src/eval/compiler/compiler.dart:134:12)
#22     main (file:///Users/maxiee/Code/dart_eval/bin/dart_eval.dart:125:36)
#23     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
#24     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:192:26)
```

I locate this line of method `if (route == 'sub') return const SubPage();` in Gallery [main.dart](https://github.com/maxiee/flutter_eval_gallery/blob/f8c63b83573a787357c5a2dc218978f3160520bf/template/lib/main.dart#L16).

Looking through the exception call stack, this occurs where [IdentifierReference.getValue L301](https://github.com/ethanblake4/dart_eval/blob/a597622c23aa2cbb0f6846ff53b377519aa11fa0/lib/src/eval/compiler/reference.dart#L301) creates the DeferredOrOffset:

```dart
if (ctx.topLevelDeclarationPositions[_decl.sourceLib]?.containsKey('$name.') ?? false) {
  offset =
      DeferredOrOffset(file: _decl.sourceLib, offset: ctx.topLevelDeclarationPositions[ctx.library]!['$name.']);
} else {
```

This is a scenario where only the offset is given, but when it actually runs here, the offset is taken out to be null. `ctx.topLevelDeclarationPositions[ctx.library]` actually returns `{}`.

The value of ctx.topLevelDeclarationPositions:

```
ctx.{0: {}, 1: {}, 2: {}, 3: {}, 4: {}, 5: {}, 6: {}, 7: {}, 8: {}, 9: {}, 10: {}, 11: {Colors._: 34184}, 12: {}, 13: {}, 14: {}, 15: {}, 16: {}, 17: {}, 18: {}, 19: {}, 20: {}, 21: {}, 22: {}, 23: {}, 24: {}, 25: {}, 26: {}, 27: {}, 28: {}, 29: {}, 30: {Curves._: 34188}, 31: {}, 32: {}, 33: {}, 34: {}, 35: {}, 36: {}, 37: {}, 38: {}, 39: {}, 40: {}, 41: {}, 42: {}, 43: {SubPage.: 34236}, 44: {StatelessWidgetPage.: 34286}, 45: {StatefulWidgetPage.: 34297, _StatefulWidgetPageState.: 34366}, 46: {ViewModel.: 34386, ChangeNotifierPage.: 34400, _ChangeNotifierPageState.: 34545}, 47: {}}
```

I notice that `_decl.sourceLib` has a value of 43 and `ctx.library` has a value of 47, 47 corresponds to a `{}` which causes the offset to be null according to `'$name.'`.

**listSync() order is related to the problem**

In bin/dart_eval.dart, after addingFiles, I print the following for the files:

```dart
// bin/dart_eval.dart L:107
addFiles(libDir);
 
// added by me
for (final f in files) {
  print(f);
}
```

files results do differ between macOS and windows:

under macOS:

```dart
File: './lib/page/sub_page.dart'
File: './lib/widget/stateless_widget_page.dart'
File: './lib/widget/stateful_widget_page.dart'
File: './lib/widget/change_notifier_page.dart'
File: './lib/main.dart'
File: './lib/widgets/widgets.dart'
```

under Windows:

```dart
File: '.\lib\main.dart'
File: '.\lib\page\sub_page.dart'
File: '.\lib\widget\change_notifier_page.dart'
File: '.\lib\widget\stateful_widget_page.dart'
File: '.\lib\widget\stateless_widget_page.dart'
File: '.\lib\widgets\widgets.dart'
```

I came up with an idea to manually adjust the files to the fixed order of macOS in windows:

```dart
final file2 = <File>[];
file2.add(files[1]);
file2.add(files[4]);
file2.add(files[3]);
file2.add(files[2]);
file2.add(files[0]);
file2.add(files[5]);
files.clear();
files.addAll(file2);
```

After manually reordering windows to be consistent with macOS, the same exception as on macOS occurs! `decl.sourceLib`, and `ctx.library`  and `ctx.topLevelDeclarationPositions` have exactly the same values as given above, and the reason for the error is the same as the call stack.

In other words, the problem can be reproduced by manually adjusting the order of files to macOS based on the template project inside flutter_eval_gallery.

**Comment by ethanblake4** [comment]
> I notice that `_decl.sourceLib` has a value of 43 and `ctx.library` has a value of 47, 47 corresponds to a `{}` which causes the offset to be null according to `'$name.'`.

Ah, I think that's the actual issue :) I believe it should just be using `_decl.sourceLib` in both places, `ctx.library` should not be used here. If you'd like to update this PR to fix that I will merge it. (Also on line 385 there is the same problem)

**Comment by maxiee** [comment]
I've updated the PR to fix both in getStaticDispatch(line 385) and in getValue

**Comment by ethanblake4** [comment]
Thanks!
